### PR TITLE
Add Ruby2.5 runtime support

### PR DESCRIPTION
**Pull request by remore**
This PR adds a new ruby:2.5 kind to allow Ruby methods like this:

```
def main(param)
  { :result => param.upcase }
end
```

## Description
This PR is quite similar to #2415, but this time I did it in Ruby. I do believe there are certain number of developers who'd love to pick Ruby as a primary language (see #2046 for an instance).

Currently Ruby runtime is tentatively hosted at [remore/openwhisk-runtime-ruby](https://github.com/remore/openwhisk-runtime-ruby), but needless to say I'd love to move/transfer this to anywhere The Apache Software Foundation manage.

## Related issue and scope
#2046 is most likely related

## My changes affect the following components
- [ ] API
- [ ] Controller
- [ ] Message Bus (e.g., Kafka)
- [ ] Loadbalancer
- [ ] Invoker
- [x] Intrinsic actions (e.g., sequences, conductors)
- [ ] Data stores (e.g., CouchDB)
- [x] Tests
- [ ] Deployment
- [ ] CLI
- [ ] General tooling
- [x] Documentation

## Types of changes
- [ ] Bug fix (generally a non-breaking change which closes an issue).
- [x] Enhancement or new feature (adds new functionality).
- [ ] Breaking change (a bug fix or enhancement which changes existing behavior).

## Checklist:

- [x] I signed an [Apache CLA](https://github.com/apache/incubator-openwhisk/blob/master/CONTRIBUTING.md).
- [x] I reviewed the [style guides](https://github.com/apache/incubator-openwhisk/wiki/Contributing:-Git-guidelines#code-readiness) and followed the recommendations (Travis CI will check :).
- [x] I added tests to cover my changes.
- [x] My changes require further changes to the documentation.
- [x] I updated the documentation where necessary.

### CODE REVIEW

**Comment by rabbah** [comment]
Thanks for this contribution!

Two comments:

1. The container tests should be in the runtime repository and we should generally remove the duplication here; @csantanapr I think we were planning to do this already?

2. Please send an email to the list about the runtime (perhaps including a small example on how to try it). This pr #3680 will actually make it possible to try using your image on dockerhub. @csantanapr are you ok with merging that now?

Following the email we can propose creating a repository to accept the runtime into a new repository.

File: actionRuntimes/ruby2.5Action/Dockerfile [hunk 1]
```diff
@@ -0,0 +1 @@
+FROM openwhisk/action-ruby-v2.5
```
Comment (rabbah) [inline actionRuntimes/ruby2.5Action/Dockerfile]: With #3680 we won’t need this (and can remove all the others like it). 

cc @csantanapr

File: docs/parameters.md [hunk 2]
```diff
@@ -25,7 +25,7 @@ This page outlines how to configure parameters when deploying packages and actio
 
 ### Passing parameters to an action at invoke time
 
-Parameters can be passed to the action when it is invoked. These examples use JavaScript but all the other languages work the same way (see documentation on [Swift actions](./actions.md#creating-swift-actions), [Python actions](./actions.mdcreating-python-actions), [Java actions](./actions.mdcreating-java-actions), [PHP actions](./actions.mdcreating-php-actions), [Docker actions](./actions.mdcreating-docker-actions) or [Go actions](./actions.mdcreating-go-actions) as appropriate for more detailed examples).
+Parameters can be passed to the action when it is invoked. These examples use JavaScript but all the other languages work the same way (see documentation on [Swift actions](./actions.md#creating-swift-actions), [Python actions](./actions.mdcreating-python-actions), [Java actions](./actions.mdcreating-java-actions), [PHP actions](./actions.mdcreating-php-actions), [Ruby actions](./actions.md#creating-ruby-actions), [Docker actions](./actions.mdcreating-docker-actions) or [Go actions](./actions.mdcreating-go-actions) as appropriate for more detailed examples).
```
Comment (rabbah) [inline docs/parameters.md]: This list will just keep getting longer. Perhaps we should provide just one reference to the index above.

File: docs/reference.md [hunk 3]
```diff
@@ -434,6 +434,14 @@ The following Composer packages are also available:
 - guzzlehttp/guzzle       v6.3.0
 - ramsey/uuid             v3.6.1
 
+## Ruby actions
+
+Ruby actions are executed using Ruby 2.5. To use this runtime, specify the `wsk` CLI parameter `--kind ruby:2.5` when creating or updating an action. This is the default when creating an action with file that has a `.rb` extension.
```
Comment (rabbah) [inline docs/reference.md]: You’ll have modified the cli then to recognize this extension. I have an alternate proposal here: https://github.com/apache/incubator-openwhisk-cli/issues/286

File: docs/reference.md [hunk 4]
```diff
@@ -434,6 +434,14 @@ The following Composer packages are also available:
...
+
+A few Ruby gems such as `mechanize` and `jwt` are available in addition to the default and bundled gems. See Dockerfile of ruby runtime image for more information.
+
+Last but not least, you can use arbitrary gem so long as you use zipped actions to package all the dependencies. doc/actions.md will explain how to do this.
```
Comment (rabbah) [inline docs/reference.md]: Link?

File: settings.gradle [hunk 5]
```diff
@@ -28,6 +28,7 @@ include 'actionRuntimes:swift3.1.1Action'
 include 'actionRuntimes:swift4.1Action'
 include 'actionRuntimes:javaAction'
 include 'actionRuntimes:php7.1Action'
+include 'actionRuntimes:ruby2.5Action'
```
Comment (rabbah) [inline settings.gradle]: @csantanapr we can also drop all of these I think.

File: actionRuntimes/ruby2.5Action/Dockerfile (hunk 1 above)
Comment (remore) [reply actionRuntimes/ruby2.5Action/Dockerfile]: Ok. I will see for how #3680 goes then.

File: docs/parameters.md (hunk 2 above)
Comment (remore) [reply docs/parameters.md]: Updated accordingly.

File: docs/reference.md (hunk 3 above)
Comment (remore) [reply docs/reference.md]: Ok. I will keep my eyes open on it.

File: docs/reference.md (hunk 4 above)
Comment (remore) [reply docs/reference.md]: Updated.

**Comment by remore** [comment]
@rabbah  Thanks for catching this up. I have just sent an email to `dev@openwhisk.incubator.apache.org` but correct me if the email address was wrong.

File: docs/parameters.md [hunk 6]
```diff
@@ -25,7 +25,7 @@ This page outlines how to configure parameters when deploying packages and actio
 
 ### Passing parameters to an action at invoke time
 
-Parameters can be passed to the action when it is invoked. These examples use JavaScript but all the other languages work the same way (see documentation on [Swift actions](./actions.md#creating-swift-actions), [Python actions](./actions.mdcreating-python-actions), [Java actions](./actions.mdcreating-java-actions), [PHP actions](./actions.mdcreating-php-actions), [Docker actions](./actions.mdcreating-docker-actions) or [Go actions](./actions.mdcreating-go-actions) as appropriate for more detailed examples).
+Parameters can be passed to the action when it is invoked. These examples use JavaScript but all the other languages work the same way (see documentation on [this page](./actions.md) as appropriate for more detailed examples).
```
Comment (rabbah) [inline docs/parameters.md]: for more details examples ... that are language specific.

Comment (remore) [reply docs/parameters.md]: Updated.

File: tests/src/test/scala/actionContainers/Ruby25ActionContainerTests.scala [hunk 7]
```diff
@@ -0,0 +1,408 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
```
Comment (rabbah) [inline tests/src/test/scala/actionContainers/Ruby25ActionContainerTests.scala]: This will also go with #3467.

File: actionRuntimes/ruby2.5Action/build.gradle [hunk 8]
```diff
@@ -0,0 +1,2 @@
+ext.dockerImageName = 'action-ruby-v2.5'
```
Comment (rabbah) [inline actionRuntimes/ruby2.5Action/build.gradle]: And this same comment as dockerfile.

File: tests/src/test/scala/actionContainers/Ruby25ActionContainerTests.scala (hunk 7 above)
Comment (remore) [reply tests/src/test/scala/actionContainers/Ruby25ActionContainerTests.scala]: Thanks. I will remove this once #3467 get merged.

File: actionRuntimes/ruby2.5Action/build.gradle (hunk 8 above)
Comment (remore) [reply actionRuntimes/ruby2.5Action/build.gradle]: Got it🙋‍

File: ansible/files/runtimes.json [hunk 9]
```diff
@@ -109,6 +109,16 @@
                     "name": "action-php-v7.1"
                 }
             }
+        ],
+        "ruby": [
+            {
+                "kind": "ruby:2.5",
+                "default": true,
+                "deprecated": false,
+                "image": {
```
Comment (rabbah) [inline ansible/files/runtimes.json]: with #3680, the image prefix and tag should be specified explicitly (to `openwhisk` and `latest` respectively).

File: tests/src/test/scala/actionContainers/Ruby25ActionContainerTests.scala (hunk 7 above)
Comment (remore) [reply tests/src/test/scala/actionContainers/Ruby25ActionContainerTests.scala]: I have removed this at 3cc287c1f67181c1d6c66ff59bbda01b913bcfa1

File: actionRuntimes/ruby2.5Action/build.gradle (hunk 8 above)
Comment (remore) [reply actionRuntimes/ruby2.5Action/build.gradle]: Same comment as [this](https://github.com/apache/incubator-openwhisk/pull/3725#discussion_r196755418) applies here

File: ansible/files/runtimes.json (hunk 9 above)
Comment (remore) [reply ansible/files/runtimes.json]: made a change as a result of #3680

Comment (remore) [reply ansible/files/runtimes.json]: changed `prefix` and `name` temporarily to workaround the build failure. (thank you @rabbah for the quick advice to fix this)

**Comment by rabbah** [comment]
@remore i squashed and rebased this to master to resolve a new conflicts with docs i updated.

@csantanapr we'll need to document the examples to be included here for system tests; should we just test a hello world here and defer the rest (including unicode) to the runtime tests?

**Comment by rabbah** [comment]
@remore @csantanapr I rebased this to pick up the latest refactoring.

File: docs/actions.md [hunk 10]
```diff
@@ -54,13 +54,14 @@ the [Docker](docker-actions.md) action or [native binary](docker-actions.md#crea
 paths more suitable. Multiple actions may be composed together to create a longer processing pipeline called a
 [sequence](#creating-action-sequences). A more advanced form of composition is described [here](conductors.md).
 
-* [JavaScript](actions-node.md)
-* [Python](actions-python.md)
+* [Go](actions-go.md)
```
Comment (rabbah) [inline docs/actions.md]: @akrabat @csantanapr I reordered these alphabetically. FYI.

Comment (akrabat) [reply docs/actions.md]: Alphabetical ordering makes sense to me.

**Comment by jthomas** [comment]
Looks like this PR is almost ready to go but the doc re-org needs the Ruby action details in a separate markdown file now? @remore are you able to do this small change and we can get it merged?
thanks for all the hard work on this.

**Comment by rabbah** [comment]
@jthomas we need to do a couple of things apart form updating this pr:
1. make sure the ruby runtime repo adopts and passes the basic runtimes test suite
2. move the runtime remote to an apache repo (requires INFRA to create the repo)
3. successfully building and tagging the ruby runtime from that repo
4. changing the runtime manifest to use openwhisk/ruby runtime instead of @remore's personal docker prefix

**Comment by remore** [comment]
> are you able to do this small change and we can get it merged?

Yes for sure, I have just made a few changes to get everything organized in the ruby-actions.md.
Thank you for shedding the light on this. @jthomas 

> make sure the ruby runtime repo adopts and passes the basic runtimes test suite

Here is the log output of test code which is executed in my laptop with the latest code today just FYI @rabbah 
```
vagrant@ubuntu-xenial:~/openwhisk-runtime-ruby$ ./gradlew :tests:test --tests *actionContainers.Ruby25ActionContainerTests*

> Task :tests:test

runtime.actionContainers.Ruby25ActionContainerTests > action-ruby-v2.5 should run a RUBY script STANDARD_OUT
    test router? false

runtime.actionContainers.Ruby25ActionContainerTests > action-ruby-v2.5 should run a RUBY script PASSED

runtime.actionContainers.Ruby25ActionContainerTests > action-ruby-v2.5 should run and report an error for script not returning a json object PASSED

runtime.actionContainers.Ruby25ActionContainerTests > action-ruby-v2.5 should run a RUBY action and handle unicode in source, input params, logs, and result PASSED

runtime.actionContainers.Ruby25ActionContainerTests > action-ruby-v2.5 should run a RUBY script and confirm expected environment variables PASSED

runtime.actionContainers.Ruby25ActionContainerTests > action-ruby-v2.5 should fail to initialize with bad code PASSED

runtime.actionContainers.Ruby25ActionContainerTests > action-ruby-v2.5 should fail to initialize with no code PASSED

runtime.actionContainers.Ruby25ActionContainerTests > action-ruby-v2.5 should return some error on action error PASSED

runtime.actionContainers.Ruby25ActionContainerTests > action-ruby-v2.5 should support application errors PASSED

runtime.actionContainers.Ruby25ActionContainerTests > action-ruby-v2.5 should fail gracefully when an action has a TypeError exception PASSED

runtime.actionContainers.Ruby25ActionContainerTests > action-ruby-v2.5 should support the documentation examples (1) PASSED

runtime.actionContainers.Ruby25ActionContainerTests > action-ruby-v2.5 should have mechanize and activesupport gems available PASSED

runtime.actionContainers.Ruby25ActionContainerTests > action-ruby-v2.5 should support large-ish actions PASSED

runtime.actionContainers.Ruby25ActionContainerTests > action-ruby-v2.5 should support zip-encoded packages PASSED

runtime.actionContainers.Ruby25ActionContainerTests > action-ruby-v2.5 should fail gracefully on invalid zip files PASSED

runtime.actionContainers.Ruby25ActionContainerTests > action-ruby-v2.5 should fail gracefully on valid zip files that are not actions PASSED

runtime.actionContainers.Ruby25ActionContainerTests > action-ruby-v2.5 should fail gracefully on valid zip files with invalid code in main.rb PASSED

runtime.actionContainers.Ruby25ActionContainerTests > action-ruby-v2.5 should support actions using non-default entry point PASSED

runtime.actionContainers.Ruby25ActionContainerTests > action-ruby-v2.5 should support zipped actions using non-default entry point PASSED


BUILD SUCCESSFUL in 48s
3 actionable tasks: 1 executed, 2 up-to-date
```

> move the runtime remote to an apache repo (requires INFRA to create the repo)

I'm willing to proceed with [these steps](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=87296917) only if I'm eligible. If so, please let me know.

**Comment by rabbah** [comment]
There some conflicts to resolve on the PR.

**Comment by rabbah** [comment]
@remore I opened 3 issues against the ruby repo that we’ll need to also address.

**Comment by remore** [comment]
FYI: I have just squashed as well to make both git tree view and log messages cleaner. (I have just realized that my commit messages were not following comment message convention here. My apologies.)

**Comment by rabbah** [comment]
Repository requested: https://issues.apache.org/jira/browse/INFRA-16764

**Comment by jthomas** [comment]
https://github.com/apache/incubator-openwhisk-runtime-ruby repo is now available.

**Comment by remore** [comment]
The first PR has been sent:
https://github.com/apache/incubator-openwhisk-runtime-ruby/pull/1

**Comment by remore** [comment]
@rabbah Thank you for going ahead about this. 

> this gets the PR ready for merging.
Can't wait to see that happen.

**Comment by rabbah** [comment]
image now published to dockerhub https://hub.docker.com/r/openwhisk/action-ruby-v2.5 (thanks @csantanapr).
